File: src/keeper/collect.py
"""Parse bounded Yahoo page content, rejecting incomplete or changed markup."""

from typing import Any

from bs4 import BeautifulSoup


def parse_draft(html: str, season: int) -> dict[str, Any]:
    soup = BeautifulSoup(html, "html.parser")
    selected = soup.select_one("#yfa-draftresults-select option[selected]")
    if selected is None or str(season) not in selected.get_text():
        raise ValueError("draft season selection not confirmed")
    result: dict[str, Any] = {}
    for tag in soup.select('[title="This player is a keeper."]'):
        row = tag.find_parent("tr")
        player = row.select_one("a.name") if row else None
        team = row.select_one("td.last") if row else None
        if player is None or team is None:
            raise ValueError("unrecognized keeper row")
        result.setdefault(team.get_text(strip=True), []).append(
            {"player": player.get_text(strip=True), "keeper": True}
        )
    if len(result) != 12 or any(len(p) != 5 for p in result.values()):
        raise ValueError("incomplete keeper board")
    return result


def parse_trades(html: str, season: int, league_id: int) -> list[dict[str, Any]]:
    soup = BeautifulSoup(html, "html.parser")
    table = soup.select_one("table.Tst-transaction-table")
    if table is None:
        raise ValueError("transaction table missing; auth or markup changed")
    rows = table.select("tr")
    if len(rows) <= 1 and "No recent transactions" in soup.get_text() and not table.select("a"):
        return []
    trades = []
    pending = []
    dates = []
    for row in rows:
        if "Traded to" not in row.get_text():
            raise ValueError("unrecognized trade row")
        cells = row.find_all("td", recursive=False)
        assets, owner = cells[-3], cells[-1]
        team = owner.select_one("a")
        stamp = owner.select_one(".F-timestamp")
        if team is None or stamp is None:
            raise ValueError("trade owner/date missing")
        received = []
        for paragraph in assets.select("p"):
            player = paragraph.select_one("a")
            position = paragraph.select_one(".F-position")
            if player is not None and position is not None:
                received.append(f"{player.get_text(strip=True)} ({position.get_text(strip=True)})")
            elif paragraph.get_text(strip=True).startswith("Round "):
                received.append(paragraph.get_text(" ", strip=True))
            else:
                raise ValueError("unrecognized trade asset")
        pending.append({"team": team.get_text(strip=True), "received": received})
        dates.append(stamp.get_text(strip=True))
        if len(pending) == 2:
            if dates[0] != dates[1]:
                raise ValueError("trade pairing date mismatch")
            trades.append(
                {"season": season, "league_id": league_id, "date": dates[0], "teams": pending}
            )
            pending, dates = [], []
    if pending or not trades:
        raise ValueError("incomplete trade page")
    return trades


def parse_collection(
    collected: dict[str, Any], command: str, season: int, league_id: int
) -> dict[str, Any] | list[dict[str, Any]]:
    from .core import trade_key

    if collected["season"] != season or collected["league_id"] != league_id:
        raise ValueError("collector scope mismatch")
    if command == "reconcile":
        return parse_draft(collected["pages"][0], season)
    trades = [
        trade for html in collected["pages"] for trade in parse_trades(html, season, league_id)
    ]
    if collected.get("complete") is not True or collected.get("transaction_count") != len(trades):
        raise ValueError("incomplete transaction collection")
    if len({trade_key(t) for t in trades}) != len(trades):
        raise ValueError("overlapping transaction pages")
    return trades

File: src/keeper/collect.mjs
// Browser transport only. Pure parsers and keeper rules live in Python.
import { access } from 'node:fs/promises';
import { pathToFileURL } from 'node:url';
import { homedir } from 'node:os';

const [command, seasonText, profile, leagueText = '5003'] = process.argv.slice(2);
const season = Number(seasonText), league = Number(leagueText);
if (!['reconcile', 'scan-trades'].includes(command) || !Number.isInteger(season) ||
    !Number.isInteger(league) || !profile) throw Error('invalid collector arguments');
await access(profile); // Never silently create an unauthenticated profile.
const { chromium } = await import(pathToFileURL(process.env.KEEPER_PLAYWRIGHT ||
  '/home/david/DrillDeck/node_modules/playwright/index.mjs').href);
const context = await chromium.launchPersistentContext(profile, {
  headless: true,
  executablePath: process.env.KEEPER_CHROMIUM ||
    `${homedir()}/.cache/ms-playwright/chromium-1217/chrome-linux64/chrome`,
});
try {
  const page = await context.newPage();
  page.setDefaultTimeout(30000);
  await page.goto('https://hockey.fantasysports.yahoo.com/hockey/5003/draftresults',
    {waitUntil: 'domcontentloaded'});
  const options = await page.locator('#yfa-draftresults-select option').evaluateAll(
    es => es.map(e => ({value:e.value, text:e.textContent})));
  const current = Number(options.find(o => o.value === 'current')?.text.match(/\d{4}/)?.[0]);
  if (!current) throw Error('Yahoo login or season metadata unavailable; re-login manually');
  const pages = [];
  let transactionCount = 0;
  if (command === 'reconcile') {
    const option = options.find(o => o.text.includes(String(season)));
    if (!option) throw Error('requested draft season is not available in dropdown');
    // Same URL parameter used by the dropdown's change listener. Yahoo's delayed
    // YUI init can leave selectOption() without an attached event handler.
    await page.goto(`https://hockey.fantasysports.yahoo.com/hockey/5003/draftresults?draft_results_period=${option.value}`,
      {waitUntil:'domcontentloaded'});
    pages.push(await page.content());
  } else {
    if (season !== current && league === 5003)
      throw Error('historical transactions require the archived --league-id');
    let url = `https://hockey.fantasysports.yahoo.com/${season === current ? '' : `${season}/`}hockey/${league}/transactions?transactionsfilter=trade`;
    const visited = new Set();
    while (url) {
      if (visited.size >= 100) throw Error('transaction pagination incomplete');
      visited.add(url);
      await page.goto(url, {waitUntil:'domcontentloaded'});
      if (new URL(page.url()).pathname !== new URL(url).pathname)
        throw Error('transaction season/league redirect; refusing mislabeled trades');
      pages.push(await page.content());
      const meta = await page.locator('body').evaluate(body => ({
        tradeCount: body.querySelectorAll('table.Tst-transaction-table .F-trade').length,
        pagers: [...body.querySelectorAll('ul.pagingnavlist')].map(pager => {
          const last = pager.querySelector('li.last');
          return {href: last?.querySelector('a')?.href || null,
                  terminal: Boolean(last?.classList.contains('F-shade'))};
        }),
      }));
      if (meta.pagers.some(p => !p.href && !p.terminal))
        throw Error('unrecognized pagination marker');
      const next = [...new Set(meta.pagers.filter(p => p.href).map(p => p.href))];
      if (next.length > 1 || (next.length && meta.pagers.some(p => p.terminal)))
        throw Error('ambiguous pagination');
      if (meta.tradeCount > 25 || (next.length && meta.tradeCount !== 25))
        throw Error('unsupported transaction page size');
      if (!next.length && meta.tradeCount === 25 && !meta.pagers.some(p => p.terminal))
        throw Error('missing terminal pagination evidence on full page');
      transactionCount += meta.tradeCount;
      url = next[0] || '';
      if (url) {
        const parsed = new URL(url, page.url());
        if (parsed.origin !== 'https://hockey.fantasysports.yahoo.com' ||
            parsed.pathname !== new URL(page.url()).pathname ||
            parsed.searchParams.get('transactionsfilter') !== 'trade')
          throw Error('unexpected transaction pagination target');
        if (visited.has(parsed.href)) throw Error('transaction pagination incomplete: cycle');
        const offset = Number(new URL(page.url()).searchParams.get('count') || 0);
        if (parsed.searchParams.get('count') !== String(offset + 25))
          throw Error('unexpected transaction pagination offset');
        url = parsed.href;
      }
    }
  }
  process.stdout.write(JSON.stringify({pages, season, league_id:league,
    ...(command === 'scan-trades' ? {complete:true, transaction_count:transactionCount} : {})}));
} finally {
  await context.close();
}
